### ISSUE TITLE
Layer streaming: measure peak VRAM at the real shape after setup() instead of predicting it

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#349]
The long-term answer to #327, filed so the constant debate has somewhere to end.

The pre-flight *predicts* peak VRAM from a fitted formula. The whole of #327's
inter-stack disagreement is a **retention** — one bf16 logits-shaped tensor still
held when the loss backward peaks — and the record is explicit that a synthetic
probe **structurally cannot see it**: it observes its own reference, not whether the
training loop that will actually run still holds the logits. That is why
`LOGITS_RETENTION_BYTES_PER_ELEMENT = 2` is charged unconditionally and why #327 is
closed as "deliberately not changed" rather than fixed.

Measurement does not have that limitation.

## Proposal

After `setup()` — at which point the model is already streamed and resident cost is
already paid — run **one** forward + backward at the configured shape and read
`torch.cuda.max_memory_allocated()`. Compare that against measured free VRAM
instead of comparing a prediction against it.

This sees the retention because it *is* the real loop, on the real shape, with the
real loss. It replaces a stack-dependent constant with a per-run measurement, which
is the same move the pre-flight already makes for the GEMM ceiling (a per-card
constant baked into the source would be a fabrication; measuring the user's own card
in that session is the only honest input) and for the disk media type.

Precedent for the cost: the pre-flight already pays a 4096³ best-of-3 GEMM probe and,
on the disk tier, a ~9 s Windows `Get-PhysicalDisk` query.

## Open questions to settle before implementing

- **Chicken-and-egg.** The probe step can itself OOM on Linux, or spill on Windows.
  Catching `torch.cuda.OutOfMemoryError` and reporting it as a refusal is fine on
  Linux; on Windows there is nothing to catch, so the WDDM case may still need the
  predicted number as a pre-gate. Probably: keep the prediction as a cheap first
  gate, and use the measurement to *unblock* configs the prediction refused —
  which composes with #347 rather than replacing it.
- **What to do with the estimator afterwards.** It should stay: it is what makes
  `soup profile` / plan-only paths work without a GPU, and it is validated by 10
  real runs.
- **Cost at a real shape.** One step at 8B NF4 is not free. Needs measuring before
  deciding whether it is default-on, default-off, or only on a predicted miss.

## Acceptance criteria

- [ ] A measured-fit path that runs one real step at the configured shape.
- [ ] Refusal/acceptance decided on the measurement, with the prediction still
      printed alongside so a divergence is visible.
- [ ] Wall-clock cost measured at a realistic shape and documented.
- [ ] Behaviour on a probe-step OOM defined and tested on Linux; Windows behaviour
      documented honestly (it does not raise).

**Comment by MakazhanAlpamys** [comment]
Partially shipped in v0.73.1 — **staying open**, because the acceptance criteria are met for `task: sft` only.

    soup train --config soup.yaml   # with training.stream_vram_probe: true

**What the measurement found.** The premise held. Measured through the real `soup train` (RTX 3050 4 GB, Windows, torch 2.5.1, SmolLM2-135M streamed bf16, batch 1), the formula under-predicts at long sequence:

| seq | predicted | real peak | ratio |
|---|---|---|---|
| 4352 | 3.282 GB | 3.036 GB | 1.081x — safe |
| 5120 | 3.844 GB | 4.118 GB | **0.934x** |
| 6144 | 4.590 GB | 5.830 GB | **0.787x — under by 21%** |

The reason the existing guard could not see it is worth recording: all ten rows of `MEASURED_VRAM_GRID` sit at seq 256 or 512. It varies *batch*. `test_never_under_predicts` therefore asserted a global-sounding guarantee from evidence that says nothing about sequence length, and has been narrowed with an assertion that fires if a longer row is ever added.

**Open questions from the issue, answered:**
- *Chicken-and-egg* — resolved as the issue guessed. The prediction stays a cheap pre-gate; with the probe on it is demoted to advice so the measurement can overrule it, capped at 4x over budget (the largest disagreement ever measured is 21%, so beyond a small multiple it is the config that is too big).
- *What to do with the estimator* — kept. It is what makes `soup profile` and plan-only paths work without a GPU.
- *Cost at a real shape* — 1.02–5.33 s (SmolLM2-135M at 1x1024 and 2x2048; Llama-3.1-8B NF4 at 1x512), and 51.72 s at a shape that does not fit. Shipped **default-off** on that basis.
- *Probe-step OOM* — returns a distinct `oom` result and refuses. A probe that ran and *raised* is a third outcome (`failed`) and also refuses, since the CUDA context may be unusable. Windows does not raise; documented as such.

**Why this stays open.** The probe runs a plain causal-LM step, which *is* the SFT step but is not a preference loss, so it is refused for `dpo`/`orpo`/`simpo`/`kto` — which is exactly where the prediction is least accurate. One matching shape was measured (probe 6.02 GB against a real DPO step's 5.30 GB, +13.5%, the same safe direction as SFT), but one point is not a validation, and measuring a preference step properly needs the TRL trainer, which does not exist yet at the point the probe runs.

Also unresolved: the **mechanism** of the long-sequence divergence. `seq**2` from the attention score matrix is the obvious candidate and the numbers do not settle it, so the formula is overrulable but not fixed.

Full record, including three readings withdrawn during the work: [`benchmarks/gate-v0.73.1-measured-vram-fit.md`](https://github.com/MakazhanAlpamys/Soup/blob/v0.73.1/benchmarks/gate-v0.73.1-measured-vram-fit.md)

**Comment by MakazhanAlpamys** [comment]
Closing — the measured-fit path is on `main` and all four acceptance boxes are covered.

**1. "A measured-fit path that runs one real step at the configured shape."** `utils/layer_stream.py:1086` `decide_measured_fit`, deciding on `torch.cuda.max_memory_allocated`. The docstring records why it is `max_memory_allocated` and not `max_memory_reserved` — reserved runs 1.08x-1.41x allocated on the measured box, so gating on it would have refused the flagship Llama-3.1-8B NF4 configuration (3.70 GB reserved against 3.45 GB free) that runs fine with `num_alloc_retries` at 0. Gating on reserved would have refused the headline config of the feature the gate protects.

**2. "Refusal/acceptance decided on the measurement, prediction printed alongside."** Both branches carry it: the accept reason ends `(predicted X GB)` and the refusal ends `(the formula predicted X GB)`. `test_measured_bytes_is_what_vramfit_reports_not_the_prediction` pins that the two cannot be confused, and the divergence is visible in either direction — `test_a_measured_peak_over_available_is_refused_even_when_predicted_to_fit` and `test_a_measured_peak_that_fits_unblocks_a_config_the_prediction_refused`. The second is the one this issue was really asking for: measurement *unblocking* a config the formula refused, which composes with #347 rather than replacing it, exactly as the open questions predicted.

**3. "Wall-clock cost measured at a realistic shape and documented."** `benchmarks/gate-v0.73.1-measured-vram-fit.md`, which the README links as the record including the three readings withdrawn during it.

**4. "Probe-step OOM defined and tested on Linux; Windows documented honestly."** `test_a_probe_oom_is_a_refusal_not_a_fallback` and `test_a_probe_that_raised_mid_step_refuses_rather_than_falling_back` pin the Linux behaviour, `test_a_failed_probe_still_honours_a_prediction_that_refused` and `test_a_failed_probe_does_not_invent_a_refusal_when_the_formula_was_happy` pin both directions of the fallback, and `test_an_unexpected_exception_still_releases_the_pinned_ram_store` covers the cleanup path an exception could otherwise skip. `layer_stream.py:1056` states the Windows case plainly — WDDM silently spills to host memory, so there is nothing to catch there.

The estimator stayed, as the issue said it should: `test_the_formula_reproduces_its_published_prediction` and `test_the_record_is_an_under_prediction_at_the_long_shapes` keep it honest, and the long-shape under-prediction is tracked separately in #395.

Green in the full suite locally (18087 passed).
